File: modules/rss/rss-test.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* $Rev$ $Date$ */

/**
 * Test RSS data conversion functions.
 */

#include <assert.h>
#include "stream.hpp"
#include "string.hpp"
#include "rss.hpp"

namespace tuscany {
namespace rss {

ostream* writer(const string& s, ostream* os) {
    (*os) << s;
    return os;
}

const string itemEntry("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<item>"
        "<title>fruit</title>"
        "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b</link>"
        "<description>"
        "<item>"
        "<name>Apple</name>"
        "<price>$2.99</price>"
        "</item>"
        "</description>"
        "</item>\n");

const string itemTextEntry("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<item>"
        "<title>fruit</title>"
        "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b</link>"
        "<description>Apple</description>"
        "</item>\n");

const string itemNoDescriptionEntry("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<item>"
        "<title>fruit</title>"
        "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b</link>"
        "</item>\n");

const string incompleteEntry("<item>\n"
        "<title>fruit</title>"
        "<description>"
        "<Item xmlns=\"http://services/\">"
        "<name xmlns=\"\">Orange</name>"
        "<price xmlns=\"\">3.55</price>"
        "</Item>"
        "</description>"
        "</item>");

const string completedEntry("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<item>"
        "<title>fruit</title>"
        "<link></link>"
        "<description>"
        "<Item xmlns=\"http://services/\">"
        "<name xmlns=\"\">Orange</name>"
        "<price xmlns=\"\">3.55</price>"
        "</Item>"
        "</description>"
        "</item>\n");

const bool testEntry() {
    {
        const list<value> i = nilListValue + element + value("item")
                + value(nilListValue + element + value("name") + value(string("Apple")))
                + value(nilListValue + element + value("price") + value(string("$2.99")));
        const list<value> a = nilListValue + (nilListValue + element + value("entry")
                + value(nilListValue + element + value("title") + value(string("fruit")))
                + value(nilListValue + element + value("id") + value(string("cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b")))
                + value(nilListValue + element + value("content") + value(i)));
        ostringstream os;
        writeRSSEntry<ostream*>(writer, &os, a);
        assert(str(os) == itemEntry);
    }
    {
        const list<value> a = nilListValue + (nilListValue + element + value("entry")
                + value(nilListValue + element + value("title") + value(string("fruit")))
                + value(nilListValue + element + value("id") + value(string("cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b")))
                + value(nilListValue + element + value("content") + value(string("Apple"))));
        ostringstream os;
        writeRSSEntry<ostream*>(writer, &os, a);
        assert(str(os) == itemTextEntry);
    }
    {
        const list<value> a = nilListValue + (nilListValue + element + value("entry")
                + value(nilListValue + element + value("title") + value(string("fruit")))
                + value(nilListValue + element + value("id") + value(string("cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b"))));
        ostringstream os;
        writeRSSEntry<ostream*>(writer, &os, a);
        assert(str(os) == itemNoDescriptionEntry);
    }
    {
        const list<value> a = content(readRSSEntry(mklist(itemEntry)));
        ostringstream os;
        writeRSSEntry<ostream*>(writer, &os, a);
        assert(str(os) == itemEntry);
    }
    {
        const list<value> a = content(readRSSEntry(mklist(itemTextEntry)));
        ostringstream os;
        writeRSSEntry<ostream*>(writer, &os, a);
        assert(str(os) == itemTextEntry);
    }
    {
        const list<value> a = content(readRSSEntry(mklist(itemNoDescriptionEntry)));
        ostringstream os;
        writeRSSEntry<ostream*>(writer, &os, a);
        assert(str(os) == itemNoDescriptionEntry);
    }
    {
        const list<value> a = content(readRSSEntry(mklist(incompleteEntry)));
        ostringstream os;
        writeRSSEntry<ostream*>(writer, &os, a);
        assert(str(os) == completedEntry);
    }
    return true;
}

const string emptyFeed("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<rss version=\"2.0\">"
        "<channel>"
        "<title>Feed</title>"
        "<link>1234</link>"
        "<description>Feed</description>"
        "</channel>"
        "</rss>\n");

const string itemFeed("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<rss version=\"2.0\">"
        "<channel>"
        "<title>Feed</title>"
        "<link>1234</link>"
        "<description>Feed</description>"
        "<item>"
        "<title>fruit</title>"
        "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b</link>"
        "<description>"
        "<item>"
        "<name>Apple</name>"
        "<price>$2.99</price>"
        "</item>"
        "</description>"
        "</item>"
        "<item>"
        "<title>fruit</title>"
        "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83c</link>"
        "<description>"
        "<item>"
        "<name>Orange</name>"
        "<price>$3.55</price>"
        "</item>"
        "</description>"
        "</item>"
        "</channel>"
        "</rss>\n");

const bool testFeed() {
    {
        const list<value> a = nilListValue + (nilListValue + element + value("feed")
                + value(nilListValue + element + value("title") + value(string("Feed")))
                + value(nilListValue + element + value("id") + value(string("1234"))));
        ostringstream os;
        writeRSSFeed<ostream*>(writer, &os, a);
        assert(str(os) == emptyFeed);
    }
    {
        const list<value> a = content(readRSSFeed(mklist(emptyFeed)));
        ostringstream os;
        writeRSSFeed<ostream*>(writer, &os, a);
        assert(str(os) == emptyFeed);
    }
    {
        const list<value> i1 = nilListValue + element + "item"
                        + (nilListValue + element + "name" + "Apple")
                        + (nilListValue + element + "price" + "$2.99");

        const list<value> i2 = nilListValue + element + "item"
                        + (nilListValue + element + "name" + "Orange")
                        + (nilListValue + element + "price" + "$3.55");

        const list<value> i = nilListValue
            + value(nilListValue + element + value("entry")
                + value(nilListValue + element + value("title") + value(string("fruit")))
                + value(nilListValue + element + value("id") + value(string("cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b")))
                + value(nilListValue + element + value("content") + value(i1)))
            + value(nilListValue + element + value("entry")
                + value(nilListValue + element + value("title") + value(string("fruit")))
                + value(nilListValue + element + value("id") + value(string("cart-53d67a61-aa5e-4e5e-8401-39edeba8b83c")))
                + value(nilListValue + element + value("content") + value(i2)));
        
        const list<value> a = nilListValue + (append<value>(nilListValue + element + value("feed")
                + value(nilListValue + element + value("title") + value(string("Feed")))
                + value(nilListValue + element + value("id") + value("1234")),
                i));

        ostringstream os;
        writeRSSFeed<ostream*>(writer, &os, a);
        assert(str(os) == itemFeed);
    }
    {
        const list<value> a = content(readRSSFeed(mklist(itemFeed)));
        ostringstream os;
        writeRSSFeed<ostream*>(writer, &os, a);
        assert(str(os) == itemFeed);
    }
    {
        const list<value> a = elementsToValues(content(readRSSFeed(mklist(itemFeed))));
        ostringstream os;
        writeRSSFeed<ostream*>(writer, &os, valuesToElements(a));
        assert(str(os) == itemFeed);
    }
    return true;
}

}
}

int main() {
    const tuscany::gc_scoped_pool p;
    tuscany::cout << "Testing..." << tuscany::endl;

    tuscany::rss::testEntry();
    tuscany::rss::testFeed();

    tuscany::cout << "OK" << tuscany::endl;

    return 0;
}

File: modules/wsgi/rss-test.py
#!/usr/bin/python
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#  
#    http://www.apache.org/licenses/LICENSE-2.0
#    
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

# Test RSS data conversion functions

import unittest
from elemutil import *
from rssutil import *

itemEntry = \
    "<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n" \
    "<item>" \
    "<title>fruit</title>" \
    "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b</link>" \
    "<description>" \
    "<item>" \
    "<name>Apple</name><price>$2.99</price>" \
    "</item>" \
    "</description>" \
    "</item>\n"

textEntry = \
    "<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n" \
    "<item>" \
    "<title>fruit</title>" \
    "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b</link>" \
    "<description>" \
    "Apple" \
    "</description>" \
    "</item>\n"

incompleteEntry = \
    "<item>" \
    "<title>fruit</title><description>" \
    "<fruit xmlns=\"http://services/\">" \
    "<name xmlns=\"\">Orange</name>" \
    "<price xmlns=\"\">3.55</price>" \
    "</fruit>" \
    "</description>" \
    "</item>"

completedEntry = \
    "<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n" \
    "<item>" \
    "<title>fruit</title>" \
    "<link />" \
    "<description>" \
    "<fruit xmlns=\"http://services/\">" \
    "<name xmlns=\"\">Orange</name>" \
    "<price xmlns=\"\">3.55</price>" \
    "</fruit>" \
    "</description>" \
    "</item>\n"

def testEntry():
    i = (element, "'item", (element, "'name", "Apple"), (element, "'price", "$2.99"))
    a = ("fruit", "cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b", i)
    s = writeRSSEntry(a);
    assert car(s) == itemEntry

    a2 = readRSSEntry((itemEntry,))
    s2 = writeRSSEntry(a2)
    assert car(s2) == itemEntry

    a3 = readRSSEntry((textEntry,))
    s3 = writeRSSEntry(a3)
    assert car(s3) == textEntry

    a4 = readRSSEntry((incompleteEntry,))
    s4 = writeRSSEntry(a4)
    assert car(s4) == completedEntry
    return True

emptyFeed = \
    "<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n" \
    "<rss version=\"2.0\">" \
    "<channel>" \
    "<title>Feed</title>" \
    "<link>1234</link>" \
    "<description>Feed</description>" \
    "</channel>" \
    "</rss>\n"

itemFeed = \
    "<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n" \
    "<rss version=\"2.0\">" \
    "<channel>" \
    "<title>Feed</title>" \
    "<link>1234</link>" \
    "<description>Feed</description>" \
    "<item>" \
    "<title>fruit</title>" \
    "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b</link>" \
    "<description>" \
    "<item>" \
    "<name>Apple</name><price>$2.99</price>" \
    "</item>" \
    "</description>" \
    "</item>" \
    "<item>" \
    "<title>fruit</title>" \
    "<link>cart-53d67a61-aa5e-4e5e-8401-39edeba8b83c</link>" \
    "<description>" \
    "<item>" \
    "<name>Orange</name><price>$3.55</price>" \
    "</item>" \
    "</description>" \
    "</item>" \
    "</channel>" \
    "</rss>\n"

def testFeed():
    s = writeRSSFeed(("Feed", "1234"))
    assert car(s) == emptyFeed

    a2 = readRSSFeed((emptyFeed,))
    s2 = writeRSSFeed(a2)
    assert car(s2) == emptyFeed

    i3 = (("fruit", "cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b",
            (element, "'item", (element, "'name", "Apple"), (element, "'price", "$2.99"))),
          ("fruit", "cart-53d67a61-aa5e-4e5e-8401-39edeba8b83c",
            (element, "'item", (element, "'name", "Orange"), (element, "'price", "$3.55"))))
    a3 = cons("Feed", cons("1234", i3))
    s3 = writeRSSFeed(a3)
    assert car(s3) == itemFeed

    i4 = (("fruit", "cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b",
            valueToElement(("'item", ("'name", "Apple"), ("'price", "$2.99")))),
          ("fruit", "cart-53d67a61-aa5e-4e5e-8401-39edeba8b83c",
            valueToElement(("'item", ("'name", "Orange"), ("'price", "$3.55")))))
    a4 = cons("Feed", cons("1234", i4))
    s4 = writeRSSFeed(a4)
    assert car(s4) == itemFeed

    a5 = readRSSFeed((itemFeed,));
    s5 = writeRSSFeed(a5);
    assert car(s5) == itemFeed

    i6 = (("fruit", "cart-53d67a61-aa5e-4e5e-8401-39edeba8b83b",
            (("'name", "Apple"), ("'price", "$2.99"))),
          ("fruit", "cart-53d67a61-aa5e-4e5e-8401-39edeba8b83c",
            (("'name", "Orange"), ("'price", "$3.55"))))
    a6 = cons("Feed", cons("1234", i6))
    s6 = writeRSSFeed(feedValuesToElements(a6))
    assert car(s6) == itemFeed

    return True

if __name__ == "__main__":
    print "Testing..."
    testEntry()
    testFeed()
    print "OK"

